gunrock/cuda/device_properties.hxx

```diff
@@ -11,19 +11,193 @@
+ * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capability
+ * @param capability Compute capability from which to get the result
+ * \return const char* architecture name or nullptr if capability is invalid
+ */
+inline constexpr const char* arch_name(compute_capability_t capability) {
+  return
+    (capability.major == 8) ?  "Ampere" :
+    (capability.major == 7 && capability.minor == 5) ? "Turing" :
+    (capability.major == 7) ?   "Volta" :
+    (capability.major == 6) ?  "Pascal" :
+    (capability.major == 5) ? "Maxwell" :
+    (capability.major == 3) ?  "Kepler" :
+                                nullptr ;
+}
+
+// Device properties retrieved from:
+// https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications
+
+/**
+ * @brief Maximum number of threads per block.
+ * \return unsigned
+ */
+inline constexpr unsigned cta_max_threads() {
+  return 1 << 10;  // 1024 threads per CTA
+}
+
+/**
+ * @brief Warp size (has always been 32, but is subject to change).
+ * \return unsigned
+ */
+inline constexpr unsigned warp_max_threads() {
+  return 1 << 5;
+}
+
+/**
+ * @brief Maximum number of resident blocks per SM.
+ * @param capability Compute capability from which to get the result
+ * \return unsigned
+ */
+inline constexpr unsigned sm_max_ctas(compute_capability_t capability) {
+  return
+    (capability >= 86) ? 16 :  // SM86+
+    (capability >= 80) ? 32 :  // SM80
+    (capability >= 75) ? 16 :  // SM75
+    (capability >= 50) ? 32 :  // SM50-SM72
+                         16 ;  // SM30-SM37
+}
+
+/**
+ * @brief Maximum number of resident threads per SM.
+ * @param capability Compute capability from which to get the result
+ * \return unsigned
+ */
+inline constexpr unsigned sm_max_threads(compute_capability_t capability) {
+  return
+    (capability >= 86) ? 1536 :  // SM86+
+    (capability >= 80) ? 2048 :  // SM80
+    (capability >= 75) ? 1024 :  // SM75
+                         2048 ;  // SM30-SM72
+}
+
+/**
+ * @brief Number of 32-bit registers per SM.
+ * @param capability Compute capability from which to get the result
+ * \return unsigned
+ */
+inline constexpr unsigned sm_registers(compute_capability_t capability) {
+  return
+    (capability >= 50) ?  64 * K :  // SM50+
+    (capability >= 37) ? 128 * K :  // SM37
+                          64 * K ;  // SM30-SM35
+}
+
+/**
+ * @brief Maximum amount of shared memory per SM.
+ * @tparam sm3XCacheConfig cudaFuncCache enum representing the shared data
+ *                         cache configuration used when called on compute
+ *                         capability 3.x
+ * @param capability       Compute capability from which to get the result
+ * \return unsigned
+ * @todo Test if this function can be resolved at compile time
+ */
+template<enum cudaFuncCache sm3XCacheConfig = cudaFuncCachePreferNone>
+inline constexpr unsigned sm_max_smem_bytes(compute_capability_t capability) {
```

Comment:
Can you rename this to `sm_max_shared_memory_bytes` to keep it consistent?

Reply:
Done